### PR TITLE
Allow for-each loops with object nodes

### PR DESCRIPTION
**Pull request by kno10**
This change allows using the much more elegant for-each constructs:
```
for(Map.Entry<String, JsonNode> enry : objectNode.fieldSet())
```
and
```
for(String field : objectNode.fieldNamesSet())
```
in addition to the current iterator-based access, which is more code intensive:
```
Iterator<String> fieldNameIterator = objectNode.fieldNames();
while(fieldNameIterator.hasNext()) {
  String fieldName = fieldNameIterator.next();
}
```
and
```
Iterator<Map.Entry<String, JsonNode>> entryIterator = objectNode.fields();
while(entryIterator.hasNext()) {
  Map.Entry<String, JsonNode> entry = entryIterator.next();
}
```

IMPORTANT: This also needs an addition to TreeNode in jackson-core, to add this interface method:
```
    /**
     * Method for accessing names of all fields for this node, iff
     * this node is an Object node. Number of field names accessible
     * will be {@link #size}.
     */
    Set<String> fieldNamesSet();
```
or the `@Override` needs to be removed from this method in `JsonNode`.

In my opinion (you may want to consider this for version 3?), the `Iterator` based methods should be replaced with `Iterable` methods completely. I.e. if API breaking changes are acceptable, I would make `fields()` return an `Iterable<Map.Entry<String, JsonNode>>` rather than an iterator, as `.fieldNames().iterator()` is exactly equivalent; and `fieldNames()` return `Iterable<String>` rather than an iterator for the same reason. Java allows foreach loops for `Iterable`, but not for `Iterator`.

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/databind/JsonNode.java [hunk 1]
```diff
@@ -770,6 +775,14 @@ public boolean hasNonNull(int index) {
         return ClassUtil.emptyIterator();
     }
 
+    /**
+     * @return Set that can be used to traverse all key/value pairs for
+     *   object nodes; empty set (no contents) for other types
+     */
+    public Set<Map.Entry<String, JsonNode>> fields() {
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/JsonNode.java]: I think this should be `fieldSet`, as there is existing `fields` method (that returns `Iterator`)?

**Comment by cowtowncoder** [comment]
Yes, I concur in all points. Addition of `Iterator` methods was sub-optimal in hindsight, compared to exposing something to support foreach.

The only change is, I think, that due to backwards compatibility concerns, method probably should not be added in `TreeNode` until 3.0. While there is no guarantee that cross-minor version components work together, we try to keep compatibility between "adjacent" minor versions at very least, and `jackson-core` is dependency for pretty much everything.
So, adding a new method in `interface` needs to happen gradually: first adding methods in implementations and then on a later minor version demote it in `interface`.
In this case 3.0 being (most likely) the next version this would work well too.
 
Actually come to think of it; instead of `Set`, I think return values should be `Iterable`s?
Or at least `Collection`s. While in practice most implementation use `Set`s, I think this is not guaranteed so it seems better to give little bit of flexibility for implementations.

File: src/main/java/com/fasterxml/jackson/databind/JsonNode.java (hunk 1 above)
Comment (kno10) [inline src/main/java/com/fasterxml/jackson/databind/JsonNode.java]: Sorry, yes. This is supposed to be `fieldSet`.

**Comment by kno10** [comment]
I chose `Set` because I modeled this after the `Map` interface of Java, which has `entrySet()`, `keySet()` etc.; but yes, I agree that `Collection` (if it allows modification!) or `Iterable` may be more appropriate.

For 3.0, I would suggest to kick the `Iterator`, as any `Iterable` or `Collection` will allow access to an `Iterator` if desired; on a new API, `node.fields().iterator()` should be the recommended replacement.

**Comment by cowtowncoder** [comment]
Yes, agreed, for 3.x can and should do bigger clean up. I'll tag this as 3.x change.

**Comment by cowtowncoder** [comment]
I think this is mostly obsoleted by #3809 (which adds `JsonNode.properties()`) -- although if alternative was needed for field/property names, could have another PR for just that part: if so, should be called `propertyNames()` to follow (new) convention of Object entries being called "properties" in all new(er) parts of API.

Closing this PR.

**Pull request closed without merging**